RPC timeout implementation

tegefaulkes opened the pull request:
### Description

This PR handles adding the timeout features to the agnostic RPC system

### Issues Fixed

* Fixes #510 
* Fixes #516 

### Tasks

- [x] 1. Update `RPCServer` to support timeouts.
- [x] 2. Update `RPCClient` to support timeouts.
- [x] 3. Update client handlers to make use of the new cancellation/timeout `ctx`.
- ~5. Update agentHandlers to make use of the new cancellation/timeout `ctx`.~ This is blocked by #512 
- [x] 6. Support communicating timeouts between `RPCClient` and `RPCServer`. support updating the timeouts based on these communicated values.
- [x] 7. Update `WebSocketClient` to take a `ctx: ContextTime` and allow it to abort the current connection.

### Final checklist

* [x] Domain specific tests
* [x] Full tests
* [x] Updated inline-comment documentation
* [x] Lint fixed
* [x] Squash and rebased
* [x] Sanity check the final build

tegefaulkes committed:
feat: `RPCServer` timeout support

[ci skip]

tegefaulkes committed:
dep: updating dependencies

* updated `@matrixai/timer` to `1.1.0`
* updated `@matrixai/async-cancellable` to `1.0.4`

[ci skip]

tegefaulkes committed:
fix: reduced some complexity with `handleStream`

[ci skip]

tegefaulkes committed:
feat: `RPCClient` now handles timeouts and abort signals

[ci skip]

tegefaulkes committed:
feat: added cancellation logic to streaming handlers

[ci skip]

tegefaulkes committed:
feat: `WebSocketClient` supports stream abortion

[ci skip]

tegefaulkes committed:
fix: propagating `ctx` to websocket stream creation

[ci skip]

tegefaulkes committed:
test: `RPCServer` test for updating timeout in middleware

[ci skip]

tegefaulkes committed:
feat: added timeout middleware to `client`

[ci skip]

tegefaulkes committed:
fix: `RPCClient` cleans up `unary` and `server` streaming calls

For duplex and client streaming it is required that the user manually cleans up streams.

[ci skip]

CMCDragonkai commented on src/PolykeyAgent.ts:
```diff
@@ -12,8 +12,8 @@ import { DB } from '@matrixai/db';
 import WebSocketServer from './websockets/WebSocketServer';
-import * as middlewareUtils from './rpc/utils/middleware';
-import * as authMiddleware from './client/utils/authenticationMiddleware';
+import * as rpcMiddlewareUtils from './rpc/utils/middleware';
```
This should be called `rpcMiddleswareUtils`. You should follow the directory structure if you're renaming. But I think we will re-org the middleware structure later.

tegefaulkes replied:
I'm confused, it is called `rpcMiddlewareUtils`, do you mean the `clientMiddleware` below it should be called `clientMiddlewareUtils`?

CMCDragonkai replied:
Oh I mean it should be `rpcUtilsMiddleware`. But actually the middleware should have separate structure later.

The name should match the directory path.

tegefaulkes replied:
This has been updated.

CMCDragonkai commented on src/bin/agent/CommandLockAll.ts:
```diff
@@ -58,7 +58,7 @@ class CommandLockAll extends CommandPolykey {
         });
         pkClient = await PolykeyClient.createPolykeyClient({
-          streamFactory: () => webSocketClient.startConnection(),
+          streamFactory: (ctx) => webSocketClient.startConnection(ctx),
```
Note that this `ctx` primarily contains only a `signal` right? Or does it also container a timer?

```
ctx.timer
ctx.signal
```

If it contains both, then we could use both to know what to do. I think the stream factory will only deal with the `signal`.

The only time it would also use the `ctx.timer` is when the timer is `ctx` is passed to down to any timed operations.

Just trying to remember how we intended to make this work.

CMCDragonkai replied:
Also note that in the js-quic this will need to be updated.

Specifically:

```
QUICClient.createQUICClient
QUICConnection.connectQUICConnection
QUICConnections.streamNew
```

tegefaulkes replied:
For the `streamFactory` we're passing a `ContextTimed`, the timer is used for timeout for the connection here since the websocket makes a new connection per stream. The signal is used to abort an active connection.

For the quic system only the signal will be used to abort an active stream. A timer could possibly be used to timeout the stream negotiation if the connection is not responding at this point but I'm unsure of the internal details here.

CMCDragonkai replied:
But when timer is exhausted, the signal is aborted. So we only need to watch the signal.

But it seems you're also saying to watch it across the lifecycle of the stream too.

CMCDragonkai replied:
See: https://github.com/MatrixAI/Polykey/pull/513#discussion_r1156712794

CMCDragonkai commented on src/bin/identities/CommandPermissions.ts:
```diff
@@ -56,7 +56,7 @@ class CommandPermissions extends CommandPolykey {
         });
         pkClient = await PolykeyClient.createPolykeyClient({
-          streamFactory: () => webSocketClient.startConnection(),
+          streamFactory: (ctx) => webSocketClient.startConnection(ctx),
```
Note that in QUIC, creating a new stream involves potentially creating a new connection and stream.

However here, you already have the `webSocketClient` and you are calling `startConnection`.

Does that mean... when using this in a QUIC situation, then you'd be using `quicConnection.streamNew`?

tegefaulkes replied:
We call whatever is creating a new stream, so yes, for the quic system we'd call the `streamNew` method. The life-cycle of the connection itself would be maintained separately here.

CMCDragonkai replied:
The problem here is:
1. We want to be able to `readable.cancel` or `writable.abort` due to timer deadline which means the stream is destroyed due to idleness.
2. You cannot just use `transformStream.readable.cancel()` without ownership, if called it results in a locked error.
3. The `pipeTo` and `pipeThrough` does support the abort signal, however attempting to use this does not propagate the original reason across the composed streams, instead it would use an abortion error.

Therefore the solution is to have a way to abort the original stream directly.

2 options:

1. Put an `cancel` method into the a derived `ReadableWritablePair` type. We can call this `RPCStream` interface.
2. Demand that the `streamFactory` returns a `PromiseCancellable<[ReadableWritablePair, cancel]>`.

Option 1 is actually the same way we did this with the `Task` type in `src/tasks` domain.

Let's go with option 1.

Create a `RPCStream` interface type with the `cancel` method.

Update all the RPC usages of `ReadableWritablePair` to instead use `RPCStream`.

Then update the transport layer stream factories to produce `RPCStream`.

That also means that `streamFactory(ctx)` now can preserve the original meaning of `timedCancellable`.

That means you can use the `timedCancellable` HOF wrapper around `streamFactory`.

Furthermore the transport layer has to understand `ctx.signal` is used for the lifetime of the asynchronous creation, not the lifetime of the stream itself.

CMCDragonkai replied:
Also because the context decorators are not a separate package. And I think there was a reason why this was the case, we cannot use the decorators, instead the HOF has to be used.

But `timedCancellable` HOF should give you back `PromiseCancellable`.

Meaning the `StreamFactory: (ctx: ContextTimed) => Promise<RPCStream<...>>`, because the wrapping happens at the RPC domain.

Note that using the `timedCancellable` allows you to configure the defaults during the wrapping.

tegefaulkes replied:
This change should work. Only concern now is that we have two timeouts, One handled by the TimedCancelable logic for creating the stream, and a separate one for the refresh.

I'm not sure how to get the keep alive timeout. Do I provide a separate option for a keep alive timeout timer separate from the CTX?

CMCDragonkai replied:
The first timeout is for acquiring the stream.

The second timeout is to keep the stream alive.

Note that when a RPC call is made, you have a timeout for the entire operation. You pass the same timer down to acquisition of the stream... then the second timeout for the entire keepalive. So from the client's perspective, these are just 2 operations that could happen.

On the server side, you only have the second timeout, the first timeout does not ever exist.

tegefaulkes replied:
As discussed. The timers will be shared for keep alive and stream timeout. So same as currently designed.

tegefaulkes replied:
This has been resolved.

CMCDragonkai commented on src/client/handlers/gestaltsGestaltList.ts:
```diff
@@ -13,14 +13,17 @@ class GestaltsGestaltListHandler extends ServerHandler<
-  > {
+  public async *handle(
+    input_,
+    connectionInfo_,
```
Why do these both have trailing underscores?

tegefaulkes replied:
They're not being used here. I may have gotten the type of underscore mixed up.

Leading means unused and trailing means alternative/internal? I'll fix them up.

CMCDragonkai replied:
So we have 3 naming conventions:

1. `this._x` means it is externally immutable, but internally mutable, this is usually combined with `get x(){ return this._x; }`.
2. `f(_a)` means un-used.
3. `a_` is alternative, because this is the most similar to `a'`, which is how haskell conventionally indicates an "updated" variable with the same meaning

tegefaulkes replied:
I wasn't aware that 1. was a convention we were using but looking at the code again I can see that this is the case. I'll fix the usage.

tegefaulkes replied:
This has been fixed.

CMCDragonkai commented on src/client/handlers/gestaltsGestaltList.ts:
```diff
@@ -13,14 +13,17 @@ class GestaltsGestaltListHandler extends ServerHandler<
       tran,
     ): AsyncGenerator<ClientRPCResponseResult<GestaltMessage>> {
       for await (const gestalt of gestaltGraph.getGestalts(tran)) {
+        if (ctx.signal.aborted) throw ctx.signal.reason;
```
I wonder if it's better to do it a bit differently.

This:

```
if (ctx.signal.aborted) throw ctx.signal.reason;
```

Should be done at the beginning of the function before any `await`.

Then it can also be checked at the end of each loop operation.

However that's not them most ideal.

The most ideal, is that the `ctx` is passed directly into the operation itself that is here would be `gestaltGraph.getGestalts(tran, ctx)`.

Note that you should also then check before every `yield`.

But it is not necessary to check on each synchronous loop.

tegefaulkes replied:
I'll make the changes.

tegefaulkes replied:
I've updated where relevant, The handlers will check for abortion just before yielding  and at the beginning of the handler.

CMCDragonkai replied:
Why not propagate the `ctx` directly into `getGestalts` too? Or is that too much work atm?

CMCDragonkai commented on src/client/handlers/gestaltsGestaltList.ts:
```diff
@@ -32,11 +35,13 @@ class GestaltsGestaltListHandler extends ServerHandler<
         const newGestalt = gestaltMessage.gestalt;
         newGestalt.identities = gestalt.identities;
         for (const [key, value] of Object.entries(gestalt.nodes)) {
+          if (ctx.signal.aborted) throw ctx.signal.reason;
```
Like here, this is inside a synchronous loop, it's not necessary to check here, since it is not possible for the signal to be aborted while inside a loop.

Remember JS is singlethreaded here, so the signal can only be aborted IF there's an async await or yield.

tegefaulkes replied:
Fixed.

CMCDragonkai commented on src/client/handlers/gestaltsGestaltList.ts:
```diff
@@ -32,11 +35,13 @@ class GestaltsGestaltListHandler extends ServerHandler<
           };
         }
         for (const keyA of Object.keys(gestalt.matrix)) {
+          if (ctx.signal.aborted) throw ctx.signal.reason;
```
Same here, remove this.

CMCDragonkai replied:
Look for any situation BEFORE any `await` or `yield`.

Ideally if that `await op()` can take the `ctx` directly, then you don't need to check it.

CMCDragonkai commented on src/client/handlers/identitiesAuthenticate.ts:
Apply the principles as in `src/client/handlers/gestaltsGestaltList.ts` regarding the checking of the context signal.

tegefaulkes replied:
Fixed.

CMCDragonkai commented on src/client/handlers/vaultsCreate.ts:
```diff
@@ -28,4 +28,4 @@ class VaultsCreatehandler extends UnaryHandler<
 }
 
-export { VaultsCreatehandler };
+export { VaultsCreateHandler };
```
If this is the only class, it should be default exported. Then you can just do `import VaultsCreateHandler from './vaultsCreate`.

Actually I think it should be `import VaultsCreateHandler from ./VaultsCreateHandler`.

tegefaulkes replied:
I intentionally avoided the default export here to enforce the name of the class when importing. Keep in mind, with default exports you can import them with any name, E.G. `import SomeRandomName from './VaultsCreate'` is valid. Since these handlers are imported in large blocks I wanted to avoid a situation where the import names could detach from the actual class name if we're not being careful.

CMCDragonkai replied:
It was just convention when there's single class files. It's only lower case names when you intend to export other things. We use this standard elsewhere.

CMCDragonkai replied:
We can deal with this later. It was because of legacy from GRPC.

CMCDragonkai commented on src/client/utils/middleware.ts:
So we have generic middleware inside `src/rpc` right? Then client specific middleware and agent specific middleware?

I reckon any given middleware should have its own file.

tegefaulkes replied:
That's what I've been doing, There are two main middlewares now, authentication and timeout. This `middleware` file just combines the two and is used as the single import for most use cases.

CMCDragonkai commented on src/client/utils/timeoutMiddleware.ts:
```diff
@@ -0,0 +1,91 @@
+ */
+function timeoutMiddlewareClient(ctx: ContextTimed) {
+  const currentTimeout = ctx.timer.delay;
+  let forwardFirst = true;
+  let reverseFirst = true;
```
What are these for? Should have comments describing the base case.

tegefaulkes replied:
We only want to process the first message for the forward and reverse path. Theses are flags for that. I'll add comments.

tegefaulkes replied:
Fixed.

CMCDragonkai commented on src/rpc/RPCClient.ts:
```diff
@@ -39,12 +42,15 @@ class RPCClient<M extends ClientManifest> {
     manifest,
     streamFactory,
     middlewareFactory = middlewareUtils.defaultClientMiddlewareWrapper(),
+    defaultTimeout = 60_000, // 1 minuet
```
Typo here.

But this should be called `xTime`.

Note `defaultTimeout`.

This is because `timeout` as a word should be reserved for an actual timeout object. At least in our code here. See `src/network/` for examples for this.

This is technically the amount of time.

Also we can specific what is `x`.

Is it `streamKeepAliveTime`?

Because remember this time is different depending on unary call vs stream call. During stream call it's a "keep alive". During unary, it's applied once for the entire unary response.

Therefore you should call it `streamKeepAliveTimeoutTime`, we have something similar in `Proxy` as `connKeepAliveTimeoutTime`.

We can be more specific here, in case in the future we have other "times".

No need to specify default as that would be obvious here.

Also note that you don't document what happens when `defaultTimeout` is just `undefined`. You might want to indicate that having undefined means there's no default timeout at all.

tegefaulkes replied:
- We have a bunch of timeout related names in the domains `rpc` and `websocket`.
- `RPCClient` has `defaultTimeout` - This is the default timeout used if no timeout was provided for a client call.
- `RPCServer` has `defaultTimeout` - This is the default used if no timeout was provided for the handler definition. and `forceEndDelay` - Is the time between a handler timing out with signal which is advisory and when the stream is forced closed.
- `WebSocketClient` has `connectionTimeout`, `pingTimeout` and `pingInterval`.
- `WebSocketServer` has `idleTimeout`, `pingInterval` and `pingTimeout`. 

Changes
- `defaultTimeout` => `streamKeepAliveTimeoutTime`
- `forceEndDelay` => `timeoutForceCloseTime`
- `connectionTimeout` => `connectionTimeoutTime`
- `pingTimeout` => `pingTimeoutTime`
- `pingInterval` => pingIntervalTime`
- `idleTimeout` => `connectionIdleTimeoutTime`

CMCDragonkai replied:
What's the difference between `connectionIdleTimeoutTime` and `pingTimeoutTime`?

CMCDragonkai replied:
Also if `connectionTimeoutTime` exists on `WebSocketClient`, is the equivalent `connectionIdleTimeoutTime` on the `WebSocketServer`? If so, why not make these the same name?

tegefaulkes replied:
This has been updated. I also clarified some of the doc comments.

CMCDragonkai commented on src/rpc/RPCClient.ts:
```diff
@@ -78,22 +86,24 @@ class RPCClient<M extends ClientManifest> {
       get: (_, method) => {
         if (typeof method === 'symbol') throw never();
         switch (this.callerTypes[method]) {
           case 'UNARY':
-            return (params) => this.unaryCaller(method, params);
+            return (params, ctx) => this.unaryCaller(method, params, ctx);
           case 'SERVER':
-            return (params) => this.serverStreamCaller(method, params);
+            return (params, ctx) =>
+              this.serverStreamCaller(method, params, ctx);
           case 'CLIENT':
-            return () => this.clientStreamCaller(method);
+            return (ctx) => this.clientStreamCaller(method, ctx);
           case 'DUPLEX':
-            return () => this.duplexStreamCaller(method);
+            return (ctx) => this.duplexStreamCaller(method, ctx);
           case 'RAW':
-            return (header) => this.rawStreamCaller(method, header);
+            return (header, ctx) => this.rawStreamCaller(method, header, ctx);
```
Can you confirm in your spec that timeout when used in streams is per-message? Specifically it should be per-chunk, since a message doesn't yet exist.

For any given chunk of data, the timeout should be reset.

tegefaulkes replied:
I'm refreshing the timeout per message by design. It's too cheap for a bad actor to hold a connection open if it refreshes per chunk.

If I wanted to, even for a unary call I could send a single byte every 50ish seconds to hold a connection open. I wouldn't have to send a valid message at all to do this so long as it's valid JSON for parsing.  I could do this for a very long time until I hit the message parsing buffer limit.

Having it per message means you'd have to at least send valid messages and conform to the protocol of the RPC method you're using.

CMCDragonkai replied:
What about raw streams?

tegefaulkes replied:
Raw streams doesn't support the refreshing. It will be up to the handlers to implement the refreshing logic. Mostly for the same reason above.

CMCDragonkai replied:
> If I wanted to, even for a unary call I could send a single byte every 50ish seconds to hold a connection open. I wouldn't have to send a valid message at all to do this so long as it's valid JSON for parsing. I could do this for a very long time until I hit the message parsing buffer limit.

This isn't really an issue.

The purpose of timeouts are is not to deal with DOS. It's to deal with idle clients/idle servers.

What you're describing is perfectly fine.

tegefaulkes replied:
Ok, I'll have to make some changes then.

CMCDragonkai replied:
Raw streams cannot have timeouts... because the transport layer is already doing this. So this is not relevant anymore. But we need comments inside explaining why.

CMCDragonkai replied:
If you are resetting the timer every time you receive a message and every time you respond. In this case resetting before the response is sent.

You have to deal with the possibility that the other side is not consuming any of the data you send.

From the client's perspective the clock starts as soon as you attempt a call. Meaning that if the server is "SLOW" then eventually it will be aborted.

From the server's perspective, if you are responding with data, the client may be SLOW. Then in that case, you want to make sure that the time spent responding is also tracked.

CMCDragonkai replied:
We should do some tests that test race conditions here.. just to see how client timer and server timer interacts with large request and large responses. Imagine the request/response messages are the size of 1 MiB.

Message size should be considered here.

What we want to prevent is memory overflow here. Even in the case of raw streams.

Can you make sure the buffer size for JSON parsing is part of the `src/config.ts` as well.

tegefaulkes replied:
Buffer limit for parsing has been added to the config.

As for testing race conditions, I don't see the need right now. The difference between the client timing itself out or the server doing it is pretty minor. If the client times itself out then the stream ends with a timeout error.  If the server does it then the client sees the connection ending early and the streams end with a different error, connection ended early I think.

I could add more informative websocket stream ending with error codes. But I don't see the need right now. I can create issues for both timeout race testing and more informative error codes for websockets ending.

CMCDragonkai replied:
> As for testing race conditions, I don't see the need right now. The difference between the client timing itself out or the server doing it is pretty minor. If the client times itself out then the stream ends with a timeout error. If the server does it then the client sees the connection ending early and the streams end with a different error, connection ended early I think.

Can you make sure there are tests covering when the client timesout VS when the server timesout. Please give the name of the test here.

> I could add more informative websocket stream ending with error codes. But I don't see the need right now. I can create issues for both timeout race testing and more informative error codes for websockets ending.

This needs to be designed along with the mapping functions use in js-quic. Can you point out where in our spec that error codes is being specified?

tegefaulkes replied:
RPCServer has two tests, if the stream fails and plenty of timeout tests.
* `'should emit stream error if input stream fails'`
* `'should emit stream error if output stream fails'`

The `RPCClient` doesn't have any tests. But it doesn't do anything complex here. if it's timing out then it cancels the stream, if the stream ends early then the stream throws. Both cases is the stream throwing an error. 

I don't expect any odd interactions happening depending on race conditions with timeouts. So I didn't write any integration tests with that in mind. 

As for the WS error codes. It's not in the spec. There is only 1 code I'm using `4001` as a generic catch all. We don't use the closing frames to convey any errors. the only time it would contain messages is if we ended the connection abruptly and we only use that information for debug messages.

We could expand this feature out but I don't think it's needed.  It's also out of scope for this PR.

CMCDragonkai replied:
Can you change to 4000 as the generic error code. 4001 and above should be specific codes. This matches the `0` being used in QUIC.

CMCDragonkai commented on src/rpc/RPCClient.ts:
```diff
@@ -133,45 +146,87 @@ class RPCClient<M extends ClientManifest> {
+      return callerInterface.readable;
+    } finally {
+      // Clean up if any problems, ignore errors if already closed
+      await writer.close().catch(() => {});
```
Is it possible other exceptions may exist? If we can identify a specific type, it would be better to check that type before discarding it, otherwise throw upwards.

tegefaulkes replied:
It's in a finally block so I wanted to avoid duplicating the error if there was one. In fact on reviewing it, i'm being dumb here, I should be cleaning up the readable if there is an error with the writable here. I'll fix it.

tegefaulkes replied:
I've fixed this up for the server streaming. Unary caller does the same thing however. 

Inside the try block we're closing the streams, these may throw errors but it's unlikely. In the finally we try the clean up but these will usually throw an error since the should already be closed in the try block. Any errors should come from inside the try block ,these are just here to double check that they closed in case of an error.

Maybe it could be done better?

CMCDragonkai replied:
Every time there is an empty catch, we must have a comment explaining why. And it should point to later parts of the code which does either handle the exception for some reason.

tegefaulkes replied:
I've added comments for the `.catch()` usage.

CMCDragonkai commented on src/rpc/RPCClient.ts:
```diff
@@ -185,31 +240,98 @@ class RPCClient<M extends ClientManifest> {
+        {
+          nodeId: IdInternal.fromBuffer<NodeId>(Buffer.alloc(32, 0)), // FIXME
+          host: '',
+          port: 0,
```
Why are these hardcoded?

tegefaulkes replied:
It's placeholder for some metadata used then creating errors. I haven't found a good way to get the information we want here. It's something I need to review and come back to.

CMCDragonkai replied:
Not all errors should require a host and port then?

tegefaulkes replied:
All remote errors should, we'd want to know where the error came from. 

The problem here is, the information is supplied from the client side, but given that the `RPCClient` is agnostic to the transport. We don't have this information within the `RPCClient`, we shouldn't have it. so I don't have a nice way of supplying this information right now. 

Two ways of solving this come to mind.
1. The error revival could be done as a middleware, the metadata can be passed to the middleware and bypass the `RPCClient` internals that way.
2. The metadata can be supplied by the server side, but we'd have a similar problem, So it would have to be part of the middleware as well I think.

CMCDragonkai replied:
We can reuse the `RPCStream` to do this:

```ts
interface RPCStream extends ReadableWritablePair {
  cancel(reason?: any) => void;
  meta?: POJO;
}
```

Then this `streamFactory` which produces the `RPCStream` interface, is the only bridge between the transport domain and the RPC domain.

That's the only place to provide any data about the stream.

tegefaulkes replied:
Metadata such as this is provided by the `RPCStream` metadata now. This is done for both the server and client.

CMCDragonkai commented on src/rpc/RPCServer.ts:
```diff
@@ -72,13 +76,17 @@ class RPCServer extends EventTarget {
     >;
     sensitive?: boolean;
+    defaultTimeout?: number;
+    forceEndDelay?: number;
```
Again here naming should end with `Time`. So it can be `endTime` or whatever.

tegefaulkes replied:
This has been fixed.

CMCDragonkai commented on src/rpc/RPCServer.ts:
```diff
@@ -100,6 +111,8 @@ class RPCServer extends EventTarget {
     middlewareFactory,
     sensitive,
+    defaultTimeout = 60_000, // 1 minuet
+    forceEndDelay = 2_000, // 2 seconds
```
Note that these defaults should be also reset inside `src/config.ts`...?

See the existing `proxyConfig`.

CMCDragonkai replied:
You should add these into `src/config.ts`.

tegefaulkes replied:
They have been added.

CMCDragonkai commented on src/rpc/RPCServer.ts:
```diff
@@ -351,56 +402,116 @@ class RPCServer extends EventTarget {
+      }
       const outputStream = handler(
-        [leadingMetadataMessage, inputStream],
+        [headerMessage.value, inputStream],
```
Did we settle on `head, tail` as terminology?

tegefaulkes replied:
If I recall, you swapped the order so I propagated that change here. I don't think we decided to change the naming though.

CMCDragonkai commented on src/rpc/errors.ts:
```diff
@@ -94,6 +94,16 @@ class ErrorPolykeyRemote<T> extends ErrorPolykey<T> {
+class ErrorRpcTimedOut<T> extends ErrorRPC<T> {
+  static description = 'RPC handler has timed out';
+  exitCode = sysexits.UNAVAILABLE;
+}
```
Should be `ErrorRPC...`.

Also note `exitCode`. If these are intended to be thrown to the CLI, then do they make sense here?

The `UNAVAILABLE` is pretty generic.

tegefaulkes replied:
Fixed. These are generic to client and server, So they could be throw on the CLI.

CMCDragonkai replied:
What do you mean thrown on the CLI?

CMCDragonkai commented on src/rpc/utils/utils.ts:
```diff
@@ -487,6 +442,5 @@ export {
   toError,
   clientInputTransformStream,
   clientOutputTransformStream,
-  extractFirstMessageTransform,
```
Was this moved somewhere?

tegefaulkes replied:
It was removed. I simplified the logic this was used in.

CMCDragonkai commented on src/websockets/WebSocketClient.ts:
```diff
@@ -42,7 +45,7 @@ class WebSocketClient {
       maxReadableStreamBytes,
       expectedNodeIds,
-      connectionTimeout,
+      connectionTimeout ?? Infinity,
```
So if we can default to `Infinity`? Perhaps then instead of `undefined`, you force it to be a number, but default it to `Infinity` as well for the RPC code.

And naming here should be consistent, again should be called `xTime`.

If it's a "keepalive" thing, we should consistently name it `xKeepAliveTimeoutTime`.

tegefaulkes replied:
This has been updated.

CMCDragonkai commented on src/websockets/WebSocketClient.ts:
```diff
@@ -368,8 +414,23 @@ class WebSocketStreamClientInternal extends WebSocketStream {
+      this.signalWritableEnd(err);
+    }
+    // Then close the websocket
+    if (!this.webSocketEnded_) {
```
What is with these trailing underscores? Don't we use `this._x` for readable externally immutable properties? Make sure you're keeping the standard so I don't have to keep reminding here.

tegefaulkes replied:
You have me doubting myself now. I just checked other parts of the code and it looks like variables used internally that share a name use a postfix `something_`. Unused variables use a prefixed `_something`.

I'll review and fix any places that have this wrong.

tegefaulkes replied:
I've fixed this up.

CMCDragonkai commented on tests/rpc/RPCServer.test.ts:
```diff
@@ -741,8 +741,341 @@ describe(`${RPCServer.name}`, () => {
-  //  - Test odd conditions for handlers, like extra messages where 1 is expected.
-  //  - Expectations can't be inside the handlers otherwise they're caught.
-  //  - get the tap transform stream working
+  test('default timeout', async () => {
```
Can we be more specific about this test description?

CMCDragonkai commented on tests/rpc/RPCServer.test.ts:
```diff
@@ -741,8 +741,341 @@ describe(`${RPCServer.name}`, () => {
+      }
+      const ctxLongProm = promise<ContextTimed>();
+      class TestMethodLongTimeout extends UnaryHandler {
+        timeout = 100;
```
Overriding is setting the property?

tegefaulkes replied:
Yes, the alternative is to set it during construction.

CMCDragonkai replied:
The `timeout` property for `abstract class Handler` should have a docblock explaining what it is used for when used with downstream classes.

tegefaulkes replied:
A comment has been added.

CMCDragonkai commented on tests/rpc/RPCServer.test.ts:
```diff
@@ -741,8 +741,341 @@ describe(`${RPCServer.name}`, () => {
+    await rpcServer.destroy();
+  });
+  testProp(
+    'Middleware can update timeout timer',
```
All test descriptions should be lowercase.

CMCDragonkai commented on tests/rpc/utils/utils.test.ts:
```diff
@@ -14,29 +12,6 @@ describe('utils tests', () => {
-  );
-
   // TODO:
   //  - Test for badly structured data
```
Arbitrary todos here. All todos should be tickable tasks on the PR!

CMCDragonkai replied:
These should be issues.

tegefaulkes replied:
Fixed.

CMCDragonkai reviewed (changes requested):
First level review. Still need to go deeper in your logic that's in `rpc` and `websockets`.

tegefaulkes committed:
fix: small fix

[ci skip]

tegefaulkes committed:
fix: server streaming caller cleans up reverse if forward has an error

[ci skip]

tegefaulkes committed:
fix: fixing up `_` usage

https://github.com/MatrixAI/Polykey/pull/513#discussion_r1155521407

[ci skip]

tegefaulkes committed:
fix: Updated abort signal usage in handlers

[ci skip]

tegefaulkes committed:
fix: added comments to unclear variables

[ci skip]

tegefaulkes committed:
syntax: improved variable names and docs

[ci skip]

tegefaulkes committed:
syntax: fixed error names

[ci skip]

tegefaulkes committed:
syntax: fixed middleware utils imports

[ci skip]

CMCDragonkai commented on src/config.ts:
```diff
@@ -102,8 +102,8 @@ const config = {
-      pingInterval: 1_000, // 1 second
-      pingTimeout: 10_000, // 10 seconds
+      pingIntervalTime: 1_000, // 1 second
+      pingTimeoutTime: 10_000, // 10 seconds
```
Let's say that `networkConfig` is generic across transport options (websocket, quic).

While `rpcConfig` is specific to `src/rpc` domain.

Then `proxyConfig` gets removed once we have fully migrated.

CMCDragonkai replied:
So the difference between having values here and the domain specific creators having default values, is that these defaults for the entire application, whereas the domains may have defaults specific to the domain.

The `PolykeyAgent` and `PolykeyClient` should correspondingly not have any default values.

CMCDragonkai commented on src/PolykeyAgent.ts:
```diff
@@ -457,13 +454,13 @@ class PolykeyAgent {
-          idleTimeout: _networkConfig.idleTimeout,
-          pingInterval: _networkConfig.pingInterval,
-          pingTimeout: _networkConfig.pingTimeout,
+          maxReadableStreamBytes: networkConfig_.maxReadBufferBytes,
```
This stuff should exist inside `src/config.ts`.

tegefaulkes replied:
Any missing config has been added and names have been updated.

CMCDragonkai commented on src/client/utils/timeoutMiddleware.ts:
```diff
@@ -0,0 +1,93 @@
+ * This adds its timeout to the reverse metadata and updates it's timeout based
+ * on the forward metadata.
+ */
+function timeoutMiddlewareServer(ctx: ContextTimed) {
```
This is going to need to be abstracted out to a common middleware location. That could be `src/rpc/middleware`, that could be `src/middleware`.

I can see here you are using `ClientRPCRequestParams` here, but this could be a generic type parameter.

```
timeoutMiddlewareServer<Req, Res> {
  ...
}
```

Furthermore the generic timeout middleware should work even for raw streams. This is why I mentioned you should be able to reset the timer upon any data chunk.

CMCDragonkai replied:
So because the transport layer already has timeout for idleness that tracks bytes being transferred, there's no need to have a timeout for raw streams.

Therefore in this scenario, this is a "RPC" level timeout which has to involve JSON because you have the `timeout` property in the params.

CMCDragonkai replied:
So that means we have 2 layers of timeout.

The bottom layer is concerned with idleness at the networking packet level.

The top layer is concerned with application execution of the handlers.

CMCDragonkai replied:
So the only thing needed here is to figure out how to abstract these generic middlewares between `client` and `agent` domains.

tegefaulkes replied:
I'll add some details to the agent migration issue. this will be addressed there.

tegefaulkes replied:
I mentioned some details in the agent migration issue #512

CMCDragonkai commented on src/rpc/RPCServer.ts:
```diff
@@ -56,12 +58,21 @@ class RPCServer extends EventTarget {
+   * value used if the handler doesn't specify its own timeout time. This timeout is advisory and only results in a
+   * signal sent to the handler. Stream is forced to end after the timeoutForceCloseTime. Defaults to 60,000
+   * milliseconds.
+   * @param obj.timeoutForceCloseTime - Time before the stream is forced to end after the initial timeout time.
+   * The stream will be forced to close after this amount of time after the initial timeout. This is a grace period for
+   * the handler to handle timeout before it is forced to end. Defaults to 2,000 milliseconds.
```
The client is tracking whether the server is idle. The server is tracking whether the handler is idle.

So these 2 properties in the `RPCServer` is really about the handler.

You want to call these:

1. `handlerTimeoutTime` - when this is exhausted we notify the handler by calling the abort controller
2. `handlerTimeoutGraceTime` - this is about if the handler is not aborting, then we force "abort" by closing the streams

The idea here is to make it clear that these times are about the handler.

In both cases, you could have `Infinity` as a value.

tegefaulkes replied:
This has been fixed.

tegefaulkes committed:
syntax: fixes to test names

[ci skip]

tegefaulkes committed:
feat: streams are now using a `RPCStream` interface that extends `ReadableWritablePair` and includes a `cancel(reason?: any)` method and generic metadata

- created `RPCStream` interface, updating code
- fixing tests
- client now includes metadata in remote errors
- `RPCServer` includes metadata in the `RPCStream` and propagates cancel/meta to middleware
- improved cancellation feature of the server `RPCStream`

[ci skip]

tegefaulkes committed:
docs: adding commentary for `.catch()` usage for ignoring errors

[ci skip]

tegefaulkes committed:
dep: updating `@streamparser/json` to `^0.0.13`

[ci skip]

tegefaulkes committed:
fix: adding and propagating config values

[ci skip]

tegefaulkes committed:
fix: test fixes

tegefaulkes committed:
fix: changed WS close error code
